ostd/src/mm/page_table/cursor/mod.rs

```diff
@@ -126,72 +168,131 @@ impl<'rcu, M: PageTableMode, E: PageTableEntryTrait, C: PagingConstsTrait> Curso
 
         const { assert!(C::NR_LEVELS as usize <= MAX_NR_LEVELS) };
 
-        let new_pt_is_tracked = if should_map_as_tracked::<M>(va.start) {
-            MapTrackingStatus::Tracked
-        } else {
-            MapTrackingStatus::Untracked
-        };
+        Ok(locking::lock_range(pt, guard, va))
+    }
 
-        Ok(locking::lock_range(pt, guard, va, new_pt_is_tracked))
+    /// Gets the current virtual address.
+    pub fn virt_addr(&self) -> Vaddr {
+        self.va
     }
 
-    /// Gets the information of the current slot.
-    pub fn query(&mut self) -> Result<PageTableItem, PageTableError> {
+    /// Queries the mapping at the current virtual address.
+    ///
+    /// If the cursor is pointing to a valid virtual address that is locked,
+    /// it will return the virtual address range and the item at that slot.
+    pub fn query(&mut self) -> Result<PagesState<C>, PageTableError> {
         if self.va >= self.barrier_va.end {
             return Err(PageTableError::InvalidVaddr(self.va));
         }
 
         let rcu_guard = self.rcu_guard;
 
         loop {
+            let cur_va = self.va;
             let level = self.level;
-            let va = self.va;
 
-            let entry = self.cur_entry();
-
-            match entry.to_ref() {
-                Child::PageTableRef(pt) => {
+            let cur_child = self.cur_entry().to_ref();
+            let item = match cur_child {
+                ChildRef::PageTable(pt) => {
                     // SAFETY: The `pt` must be locked and no other guards exist.
                     let guard = unsafe { pt.make_guard_unchecked(rcu_guard) };
                     self.push_level(guard);
                     continue;
                 }
-                Child::PageTable(_) => {
-                    unreachable!();
+                ChildRef::None => None,
+                ChildRef::Frame(pa, ch_level, prop) => {
+                    debug_assert_eq!(ch_level, level);
+
+                    // SAFETY: It must be mapped into the page table.
+                    let item = unsafe { C::item_from_raw(pa, level, prop) };
+                    // Clone a copy so that the page table still has one ownership.
+                    // TODO: Provide a `PageTableItemRef` to reduce copies.
+                    let _ = ManuallyDrop::new(item.clone());
```

Comment:
This can be unsound if `item.clone()` panics. The unwinding process will drop the page table item while the actual item is still in the page table.

We should write something like:
```rust

                    // SAFETY: It must be mapped into the page table.
                    let item = ManuallyDrop::new(unsafe { C::item_from_raw(pa, level, prop) });
                    // TODO: Provide a `PageTableItemRef` to reduce copies.
                    Some((*item).clone())
```